cpp/src/arrow/compute/kernels/row_encoder_internal.h

```diff
@@ -232,6 +233,223 @@ struct VarLengthKeyEncoder : KeyEncoder {
   std::shared_ptr<DataType> type_;
 };
 
+template <typename T, typename ValidFunc, typename NullFunc>
+static void IterateListArray(const ArraySpan& arr, ValidFunc&& valid_func,
+                      NullFunc&& null_func) {
```

Comment:
```suggestion
void IterateListArray(const ArraySpan& arr, ValidFunc&& valid_func,
                      NullFunc&& null_func) {
```

this function doesn't need to be marked static, as its a template

Reply:
and bump on adding a comment to document that `valid_func` is invoked with.